In [5]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

## FFS

In [6]:
ffs_df = pd.read_csv("../static/data/sanfrancisco/ffs_temp.csv")
ffs_df

0
0     43.44
1     30.41
2     30.41
3     26.07
4     30.41
...     ...
4463  30.41
4464  30.41
4465  30.41
4466  21.72
4467  30.41

[4468 rows x 1 columns]

# 도로별 TCI

- speed(20080517~20080610).csv
    - 행: 5분 간격 시계열
        - 7488개
    - 열: SanFrancisco_edges의 행과 매핑
        - 4468개 (확인필요)
    - 값: 속도(km/h)

In [7]:
start_date = datetime.strptime("2008-05-17", "%Y-%m-%d")

In [8]:
df = pd.read_csv("../static/data/sanfrancisco/speed(20080517~20080610).csv").transpose()
df

0          1          2          3          4          5     \
0           NaN        NaN        NaN        NaN        NaN        NaN   
1           NaN        NaN        NaN        NaN        NaN        NaN   
2           NaN        NaN        NaN        NaN        NaN        NaN   
3     29.900222        NaN        NaN        NaN        NaN        NaN   
4           NaN        NaN  11.791672        NaN        NaN        NaN   
...         ...        ...        ...        ...        ...        ...   
4463  15.298527  23.775652        NaN        NaN        NaN  28.831172   
4464  23.128912  32.981365        NaN  12.623639        NaN  22.605815   
4465  22.477395        NaN        NaN  31.578574        NaN        NaN   
4466  22.692649  24.929268  20.812846  11.526812  20.371446  21.200411   
4467        NaN        NaN        NaN        NaN        NaN        NaN   

           6          7     8          9     ...  7478  7479  7480  7481  \
0           NaN        NaN   NaN        NaN  ...   NaN   NaN   NaN   NaN   
1           NaN        NaN   NaN        NaN  ...   NaN   NaN   NaN   NaN   
2           NaN        NaN   NaN        NaN  ...   NaN   NaN   NaN   NaN   
3     55.781078        NaN   NaN        NaN  ...   NaN   NaN   NaN   NaN   
4           NaN        NaN   NaN        NaN  ...   NaN   NaN   NaN   NaN   
...         ...        ...   ...        ...  ...   ...   ...   ...   ...   
4463        NaN  27.093892   NaN  33.085296  ...   NaN   NaN   NaN   NaN   
4464  24.010962        NaN   NaN        NaN  ...   NaN   NaN   NaN   NaN   
4465        NaN        NaN   NaN  14.339226  ...   NaN   NaN   NaN   NaN   
4466  28.826978  32.304043   NaN  16.069701  ...   NaN   NaN   NaN   NaN   
4467        NaN        NaN   NaN  29.358613  ...   NaN   NaN   NaN   NaN   

      7482  7483  7484  7485  7486  7487  
0      NaN   NaN   NaN   NaN   NaN   NaN  
1      NaN   NaN   NaN   NaN   NaN   NaN  
2      NaN   NaN   NaN   NaN   NaN   NaN  
3      NaN   NaN   NaN   NaN   NaN   NaN  
4      NaN   NaN   NaN   NaN   NaN   NaN  
...    ...   ...   ...   ...   ...   ...  
4463   NaN   NaN   NaN   NaN   NaN   NaN  
4464   NaN   NaN   NaN   NaN   NaN   NaN  
4465   NaN   NaN   NaN   NaN   NaN   NaN  
4466   NaN   NaN   NaN   NaN   NaN   NaN  
4467   NaN   NaN   NaN   NaN   NaN   NaN  

[4468 rows x 7488 columns]

In [15]:
# 날짜 단위 데이터프레임 생성
date_df = pd.DataFrame()
date_df.index = df[0].index
for i in range(0, len(df.columns), 12*24):
    cols = df.columns[i:i+12*24]
    sums = df[cols].sum(axis=1) / len(cols)
    colname = start_date + timedelta(days=int(i/(12*24)))

    date_df = pd.concat((date_df, sums.rename(colname)), axis=1)

date_df

2008-05-17  2008-05-18  2008-05-19  2008-05-20  2008-05-21  2008-05-22  \
0       5.720810         0.0         0.0         0.0         0.0         0.0   
1       0.436973         0.0         0.0         0.0         0.0         0.0   
2      11.320183         0.0         0.0         0.0         0.0         0.0   
3      17.051171         0.0         0.0         0.0         0.0         0.0   
4       0.150101         0.0         0.0         0.0         0.0         0.0   
...          ...         ...         ...         ...         ...         ...   
4463   14.614391         0.0         0.0         0.0         0.0         0.0   
4464   18.612207         0.0         0.0         0.0         0.0         0.0   
4465    6.242184         0.0         0.0         0.0         0.0         0.0   
4466   17.170523         0.0         0.0         0.0         0.0         0.0   
4467    4.177308         0.0         0.0         0.0         0.0         0.0   

      2008-05-23  2008-05-24  2008-05-25  2008-05-26  ...  2008-06-02  \
0            0.0         0.0         0.0         0.0  ...         0.0   
1            0.0         0.0         0.0         0.0  ...         0.0   
2            0.0         0.0         0.0         0.0  ...         0.0   
3            0.0         0.0         0.0         0.0  ...         0.0   
4            0.0         0.0         0.0         0.0  ...         0.0   
...          ...         ...         ...         ...  ...         ...   
4463         0.0         0.0         0.0         0.0  ...         0.0   
4464         0.0         0.0         0.0         0.0  ...         0.0   
4465         0.0         0.0         0.0         0.0  ...         0.0   
4466         0.0         0.0         0.0         0.0  ...         0.0   
4467         0.0         0.0         0.0         0.0  ...         0.0   

      2008-06-03  2008-06-04  2008-06-05  2008-06-06  2008-06-07  2008-06-08  \
0            0.0         0.0         0.0         0.0         0.0         0.0   
1            0.0         0.0         0.0         0.0         0.0         0.0   
2            0.0         0.0         0.0         0.0         0.0         0.0   
3            0.0         0.0         0.0         0.0         0.0         0.0   
4            0.0         0.0         0.0         0.0         0.0         0.0   
...          ...         ...         ...         ...         ...         ...   
4463         0.0         0.0         0.0         0.0         0.0         0.0   
4464         0.0         0.0         0.0         0.0         0.0         0.0   
4465         0.0         0.0         0.0         0.0         0.0         0.0   
4466         0.0         0.0         0.0         0.0         0.0         0.0   
4467         0.0         0.0         0.0         0.0         0.0         0.0   

      2008-06-09  2008-06-10  2008-06-11  
0            0.0         0.0         0.0  
1            0.0         0.0         0.0  
2            0.0         0.0         0.0  
3            0.0         0.0         0.0  
4            0.0         0.0         0.0  
...          ...         ...         ...  
4463         0.0         0.0         0.0  
4464         0.0         0.0         0.0  
4465         0.0         0.0         0.0  
4466         0.0         0.0         0.0  
4467         0.0         0.0         0.0  

[4468 rows x 26 columns]

In [10]:
# 시간 단위 데이터프레임 생성
time_df = pd.DataFrame()
time_df.index = df[0].index
for i in range(0, len(df.columns), 12):
    cols = df.columns[i:i+12]
    sums = df[cols].sum(axis=1) / len(cols)
    colname = start_date + timedelta(hours=int(i/12))

    time_df = pd.concat((time_df, sums.rename(colname)), axis=1)

time_df

2008-05-17 00:00:00  2008-05-17 01:00:00  2008-05-17 02:00:00  \
0                0.000000             2.584118             7.222526   
1                0.000000             0.000000             0.000000   
2                0.000000             0.000000             1.705614   
3                7.140108             6.234169            18.517561   
4                0.982639             0.000000             0.000000   
...                   ...                  ...                  ...   
4463            10.673712             5.079464             8.350477   
4464            12.004531            16.808261            19.337358   
4465             8.012721             7.772032             7.959166   
4466            17.676767            21.417416            25.090732   
4467             7.512566             0.000000             0.000000   

      2008-05-17 03:00:00  2008-05-17 04:00:00  2008-05-17 05:00:00  \
0                1.919870            10.446219             7.550453   
1                0.000000             0.000000             0.428998   
2                5.619383            14.396189            19.752928   
3               17.942391            25.088754            20.767107   
4                0.000000             0.000000             0.000000   
...                   ...                  ...                  ...   
4463            13.162073            16.484508            14.510248   
4464            20.362049            14.664335            20.061740   
4465             5.348505             1.705395             2.703191   
4466            15.191752            19.641727            15.465193   
4467             0.000000            14.703550             4.256012   

      2008-05-17 06:00:00  2008-05-17 07:00:00  2008-05-17 08:00:00  \
0                5.500985             5.173004             1.211853   
1                0.000000             0.000000             3.460808   
2               35.207808            26.497608            24.053215   
3               21.398349            12.879996            21.077729   
4                0.000000             0.000000             0.000000   
...                   ...                  ...                  ...   
4463            16.854527            16.363545            16.724472   
4464            22.577356            26.088548            12.770595   
4465             7.664127             4.549895             4.108567   
4466            19.898443            10.820372             9.694335   
4467             7.898142             9.487019            10.599492   

      2008-05-17 09:00:00  ...  2008-06-11 14:00:00  2008-06-11 15:00:00  \
0                4.523758  ...                  0.0                  0.0   
1                3.427946  ...                  0.0                  0.0   
2                9.457253  ...                  0.0                  0.0   
3               18.792296  ...                  0.0                  0.0   
4                0.000000  ...                  0.0                  0.0   
...                   ...  ...                  ...                  ...   
4463            11.445172  ...                  0.0                  0.0   
4464            11.867254  ...                  0.0                  0.0   
4465             0.835730  ...                  0.0                  0.0   
4466            16.641712  ...                  0.0                  0.0   
4467            10.993040  ...                  0.0                  0.0   

      2008-06-11 16:00:00  2008-06-11 17:00:00  2008-06-11 18:00:00  \
0                     0.0                  0.0                  0.0   
1                     0.0                  0.0                  0.0   
2                     0.0                  0.0                  0.0   
3                     0.0                  0.0                  0.0   
4                     0.0                  0.0                  0.0   
...                   ...                  ...                  ...   
4463                  0.0            

In [21]:
ffs_date_df = pd.DataFrame(columns=date_df.columns)
for c in ffs_date_df.columns:
    ffs_date_df[c] = ffs_df['0']
tci_date_df = pd.DataFrame(date_df.values / ffs_date_df.values, columns=date_df.columns).transpose()
tci_date_df.index.name = "Date"
tci_date_df

0         1         2         3         4     5     6     \
Date                                                                       
2008-05-17  0.131695  0.014369  0.372252  0.654053  0.004936   0.0   0.0   
2008-05-18  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-19  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-20  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-21  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-22  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-23  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-24  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-25  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-26  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-27  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-28  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-29  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-30  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-05-31  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-01  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-02  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-03  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-04  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-05  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-06  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-07  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-08  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-09  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-10  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-11  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   0.0   

                7         8         9     ...      4458      4459      4460  \
Date                                      ...                                 
2008-05-17  0.027932  0.051887  0.392219  ...  0.052034  0.465805  0.152906   
2008-05-18  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-19  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-20  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-21  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-22  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-23  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-24  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-25  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-26  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-27  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-28  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-29  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-30  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-05-31  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-01  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-02  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-03  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-04  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-05  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-06  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2008-06-07  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.0

In [28]:
ffs_time_df = pd.DataFrame(columns=time_df.columns)
for c in ffs_time_df.columns:
    ffs_time_df[c] = ffs_df['0']
tci_time_df = pd.DataFrame(time_df.values / ffs_time_df.values, columns=time_df.columns).transpose()
tci_time_df.index.name = "Time"
tci_time_df

0     1         2         3         4     5     6     \
Time                                                                            
2008-05-17 00:00:00  0.000000   0.0  0.000000  0.273882  0.032313   0.0   0.0   
2008-05-17 01:00:00  0.059487   0.0  0.000000  0.239132  0.000000   0.0   0.0   
2008-05-17 02:00:00  0.166264   0.0  0.056087  0.710302  0.000000   0.0   0.0   
2008-05-17 03:00:00  0.044196   0.0  0.184787  0.688239  0.000000   0.0   0.0   
2008-05-17 04:00:00  0.240475   0.0  0.473403  0.962361  0.000000   0.0   0.0   
...                       ...   ...       ...       ...       ...   ...   ...   
2008-06-11 19:00:00  0.000000   0.0  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-11 20:00:00  0.000000   0.0  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-11 21:00:00  0.000000   0.0  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-11 22:00:00  0.000000   0.0  0.000000  0.000000  0.000000   0.0   0.0   
2008-06-11 23:00:00  0.000000   0.0  0.000000  0.000000  0.000000   0.0   0.0   

                         7     8         9     ...      4458      4459  \
Time                                           ...                       
2008-05-17 00:00:00  0.104055   0.0  0.070457  ...  0.057668  0.501166   
2008-05-17 01:00:00  0.000000   0.0  0.000000  ...  0.081738  0.488823   
2008-05-17 02:00:00  0.000000   0.0  0.000000  ...  0.000000  0.387889   
2008-05-17 03:00:00  0.000000   0.0  0.302319  ...  0.000000  0.707234   
2008-05-17 04:00:00  0.000000   0.0  1.445845  ...  0.032155  0.311203   
...                       ...   ...       ...  ...       ...       ...   
2008-06-11 19:00:00  0.000000   0.0  0.000000  ...  0.000000  0.000000   
2008-06-11 20:00:00  0.000000   0.0  0.000000  ...  0.000000  0.000000   
2008-06-11 21:00:00  0.000000   0.0  0.000000  ...  0.000000  0.000000   
2008-06-11 22:00:00  0.000000   0.0  0.000000  ...  0.000000  0.000000   
2008-06-11 23:00:00  0.000000   0.0  0.000000  ...  0.000000  0.000000   

                         4460      4461      4462      4463      4464  \
Time                                                                    
2008-05-17 00:00:00  0.315431  0.311381  0.449260  0.350993  0.394756   
2008-05-17 01:00:00  0.000000  0.340231  0.106261  0.167033  0.552722   
2008-05-17 02:00:00  0.000000  0.511969  0.380087  0.274596  0.635888   
2008-05-17 03:00:00  0.000000  0.452620  0.154183  0.432821  0.669584   
2008-05-17 04:00:00  0.000000  0.272939  0.544169  0.542075  0.482221   
...                       ...       ...       ...       ...       ...   
2008-06-11 19:00:00  0.000000  0.000000  0.000000  0.000000  0.000000   
2008-06-11 20:00:00  0.000000  0.000000  0.000000  0.000000  0.000000   
2008-06-11 21:00:00  0.000000  0.000000  0.000000  0.000000  0.000000   
2008-06-11 22:00:00  0.000000  0.000000  0.000000  0.000000  0.000000   
2008-06-11 23:00:00  0.000000  0.000000  0.000000  0.000000  0.000000   

                         4465      4466      4467  
Time                                               
2008-05-17 00:00:00  0.263490  0.813847  0.247043  
2008-05-17 01:00:00  0.255575  0.986069  0.000000  
2008-05-17 02:00:00  0.261729  1.155190  0.000000  
2008-05-17 03:00:00  0.175880  0.699436  0.000000  
2008-05-17 04:00:00  0.056080  0.904315  0.483510  
...                       ...       ...       ...  
2008-06-11 19:00:00  0.000000  0.000000  0.000000  
2008-06-11 20:00:00  0.000000  0.000000  0.000000  
2008-06-11 21:00:00  0.000000  0.000000  0.000000  
2008-06-11 22:00:00  0.000000  0.000000  0.000000  
2008-06-11 23:00:00  0.000000  0.000000  0.000000  

[624 rows x 4468 columns]

In [29]:
tci_date_df.to_csv("../static/data/sanfrancisco/road_tci_date.csv")
tci_time_df.to_csv("../static/data/sanfrancisco/road_tci_time.csv")

# 전체 TCI

In [34]:
city_tci_date_df = pd.DataFrame(tci_date_df.mean(axis=1), columns=["Actual"])
city_tci_date_df

Actual
Date                
2008-05-17  0.536773
2008-05-18  0.000000
2008-05-19  0.000000
2008-05-20  0.000000
2008-05-21  0.000000
2008-05-22  0.000000
2008-05-23  0.000000
2008-05-24  0.000000
2008-05-25  0.000000
2008-05-26  0.000000
2008-05-27  0.000000
2008-05-28  0.000000
2008-05-29  0.000000
2008-05-30  0.000000
2008-05-31  0.000000
2008-06-01  0.000000
2008-06-02  0.000000
2008-06-03  0.000000
2008-06-04  0.000000
2008-06-05  0.000000
2008-06-06  0.000000
2008-06-07  0.000000
2008-06-08  0.000000
2008-06-09  0.000000
2008-06-10  0.000000
2008-06-11  0.000000

In [35]:
city_tci_time_df = pd.DataFrame(tci_time_df.mean(axis=1), columns=["Actual"])
city_tci_time_df

Actual
Time                         
2008-05-17 00:00:00  0.412000
2008-05-17 01:00:00  0.405977
2008-05-17 02:00:00  0.426277
2008-05-17 03:00:00  0.482924
2008-05-17 04:00:00  0.514807
...                       ...
2008-06-11 19:00:00  0.000000
2008-06-11 20:00:00  0.000000
2008-06-11 21:00:00  0.000000
2008-06-11 22:00:00  0.000000
2008-06-11 23:00:00  0.000000

[624 rows x 1 columns]

In [37]:
city_tci_date_df.to_csv("../static/data/sanfrancisco/city_tci_date.csv")
city_tci_time_df.to_csv("../static/data/sanfrancisco/city_tci_time.csv")